airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java

```diff
@@ -0,0 +1,224 @@
+  private final CSVPrinter csvPrinter;
+  private final String tmpTableName;
+
+  public RedshiftCopier(
+                        String runFolder,
+                        SyncMode syncMode,
+                        String schema,
+                        String streamName,
+                        AmazonS3 client,
+                        JdbcDatabase redshiftDb,
+                        String s3KeyId,
+                        String s3key,
+                        String s3Region)
+      throws IOException {
+    this(RedshiftCopyDestination.DEFAULT_AIRBYTE_STAGING_S3_BUCKET, runFolder, syncMode, schema, streamName, client, redshiftDb, s3KeyId, s3key,
+        s3Region);
+  }
+
+  @VisibleForTesting
+  public RedshiftCopier(
+                        String s3BucketName,
+                        String runFolder,
+                        SyncMode syncMode,
+                        String schema,
+                        String streamName,
+                        AmazonS3 client,
+                        JdbcDatabase redshiftDb,
+                        String s3KeyId,
+                        String s3key,
+                        String s3Region)
+      throws IOException {
+    this.s3BucketName = s3BucketName;
+    this.runFolder = runFolder;
+    this.syncMode = syncMode;
+    this.schemaName = schema;
+    this.streamName = streamName;
+    this.s3Client = client;
+    this.redshiftDb = redshiftDb;
+    this.s3KeyId = s3KeyId;
+    this.s3Key = s3key;
+    this.s3Region = s3Region;
+
+    this.tmpTableName = NAMING_RESOLVER.getTmpTableName(streamName);
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
+    // have support for streaming multipart uploads;
+    // The alternative is first writing the entire output to disk before loading into S3. This is not
+    // feasible with large tables.
+    // Data is chunked into parts. A part is sent off to a queue to be uploaded once it has reached it's
+    // configured part size.
+    // Memory consumption is queue capacity * part size = 10 * 10 = 100 MB at current configurations.
+    this.multipartUploadManager =
+        new StreamTransferManager(s3BucketName, getPath(runFolder, streamName), client)
+            .numUploadThreads(DEFAULT_UPLOAD_THREADS)
+            .queueCapacity(DEFAULT_QUEUE_CAPACITY)
+            .partSize(PART_SIZE_MB);
+    // We only need one output stream as we only have one input stream. This is reasonably performant.
+    // See the above comment.
+    this.outputStream = multipartUploadManager.getMultiPartOutputStreams().get(0);
+
+    var writer = new PrintWriter(outputStream, true, StandardCharsets.UTF_8);
+    this.csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT);
+  }
+
+  public static void closeAsOneTransaction(List<RedshiftCopier> copiers, boolean hasFailed, JdbcDatabase redshiftDb) throws Exception {
+    try {
+      StringBuilder mergeCopiersToFinalTableQuery = new StringBuilder();
```

Comment:
@cgardens this was the cleanest way I could find of doing this - basically the same pattern as the `JdbcBufferedConsumerFactory`. If I understand correctly, the `executeTransaction` call on L145 will automatically rollback any inserts into the final destination tables if inserts run into any errors.

The tmp table and the S3 files will be cleaned up by the finally block.